### PR TITLE
Rewrite pool, to be a little bit simpler.

### PR DESCRIPTION
**Pull request by fengttt**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [x] Code Refactoring

## Which issue(s) this PR fixes:

issue #

## What this PR does / why we need it:

Rewrite pool code, without append and closure.    Easier to follow and (probably) more efficient.
Fixed a leak.

### CODE REVIEW

**Comment by fengttt** [comment]
The pool itself is still trouble some.    It used a T and copy T out.   This pretty much says T must be small (slice is OK) but this is easily abused.

Slice returned can be manipulated (append).    The semantics is also hard to reason, but seems OK for slices.

File: pkg/fileservice/pool_test.go [hunk 1]
```diff
@@ -53,7 +53,7 @@ func BenchmarkBytesPool(b *testing.B) {
 	for i := 0; i < b.N; i++ {
 		var v any
 		put := pool.Get(&v)
-		put()
+		defer put.Put()
```
Comment (reusee) [inline pkg/fileservice/pool_test.go:56]: Should not defer, or the pool will drain and always go the slow path.

**Review by reusee — CHANGES_REQUESTED** [review]

**Comment by reusee** [comment]
> The pool itself is still trouble some. It used a T and copy T out. This pretty much says T must be small (slice is OK) but this is easily abused.
> 
> Slice returned can be manipulated (append). The semantics is also hard to reason, but seems OK for slices.

It's the same as `sync.Pool`, callers are expected to return pointers to make allocation minimized.
With generics, we can force it to be pointers, I will send a PR.

If the return value is a slice, there is no need to reset since Get always returns the original slice.
For pointers, a reset function is required.

**Comment by reusee** [comment]
@fengttt PR sent https://github.com/matrixorigin/matrixone/pull/9415

**Comment by fengttt** [comment]
not sure enforce ptr is tbe right thing.   yet another indirection means another allocation.    i think the copy semantics is fine for what we have now, but we need to be careful, and when we need cache big obj should have considered mpool.

**Comment by reusee** [comment]
> not sure enforce ptr is tbe right thing. yet another indirection means another allocation. i think the copy semantics is fine for what we have now, but we need to be careful, and when we need cache big obj should have considered mpool.

Indirection will be on the call side and that will likely be allocated on the stack so it's fine.

**Review by reusee — APPROVED** [review]

**Pull request merged**